sound/soc/sof/topology.c

```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 	.bytes_ext_ops_count = ARRAY_SIZE(sof_dspless_bytes_ext_ops),
 };
 
+static const struct card_tplg_ops card_ops[] = {
+};
+
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	const struct firmware *fw;
+	char **tplg_files;
+	int tplg_num = 0;
 	int ret;
+	int i;
 
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
+	for (i = 0; i < ARRAY_SIZE(card_ops); i++) {
+		if (!strcmp(scomp->card->name, card_ops[i].card_name)) {
+			ret = card_ops[i].get_tplg_files(scomp, file, &tplg_files);
+			if (ret < 0)
+				return ret;
+			tplg_num = ret;
+			break;
+		}
+	}
+	if (!tplg_num)
+		goto legacy_tplg;
+
+	for (i = 0; i < tplg_num; i++) {
+		dev_dbg(scomp->dev, "Requesting %d %s\n", i, tplg_files[i]);
+		ret = firmware_request_nowarn(&fw, tplg_files[i], scomp->dev);
+		if (ret < 0) {
+			if (i == 0) {
+				dev_dbg(scomp->dev, "Fail back to %s\n", file);
+				kfree(tplg_files[i]);
+				goto legacy_tplg;
+			}
+
+			dev_err(scomp->dev, "tplg request firmware %s failed err: %d\n",
+				tplg_files[i], ret);
+			kfree(tplg_files[i]);
+			goto out;
+		}
+		kfree(tplg_files[i]);
+
+		/* set complete = sof_complete if it is the last topology */
+		if (i == tplg_num - 1)
+			sof_tplg_ops.complete = sof_complete;
+
+		if (sdev->dspless_mode_selected)
+			ret = snd_soc_tplg_component_load(scomp, &sof_dspless_tplg_ops, fw);
+		else
+			ret = snd_soc_tplg_component_load(scomp, &sof_tplg_ops, fw);
+
+		release_firmware(fw);
+
+		if (ret < 0) {
+			dev_err(scomp->dev, "tplg component load failed %d\n", ret);
```

Comment:
the return value will be printed anyway, can we print the name of the topology that failed here?

Reply:
Yes, done